### ISSUE TITLE
Formula Support

### ISSUE THREAD

**Issue by HackintoshMan** [FlyingDiver/Indigo-Shims#1]
It would be nice to have dynamic formulas to manipulate data after extracting it from a payload. Using the `eval` method in Python might be of use since it allows for pretty low level access to data.

```
# Apply Formulas
if dev.pluginProps[u"temperatureFormula"]:
    x = temperature
    temperature = eval(dev.pluginProps[u"temperatureFormula"])
if dev.pluginProps[u"humidityFormula"]:
    x = humidity
    humidity = eval(dev.pluginProps[u"humidityFormula"])
if dev.pluginProps[u"pressureFormula"]:
    x = pressure
    pressure = eval(dev.pluginProps[u"pressureFormula"])
```
Here is an example of a similar thing I have done using eval where I store the value as `x` and it allows the user to use `x` when trying to manipulate a value. You can do things like `x + 12` or even `(x / 33.8638) + 26` for offsets and unit conversion.

I think eval can be pretty powerful and the user has access to a lot of data, so this might be something to consider. But, this would also allow access to data within a dictionary, so formulas could be applied to any data stored in any format within a JSON payload.

This could be useful for adding offsets to inaccurate sensors and for change units (say from Fahrenheit to Celsius or mb to inHg).

**Comment by FlyingDiver** [comment]
Does there have to be multiple formulas per device?  Or just one for the device value and ignore the rest of the states.  If you have different devices for each of Temp/Humidity/Pressure, then you only need one formula in each device, right?

**Comment by FlyingDiver** [comment]
```
if dev.pluginProps[u"pressureFormula"]:
    x = value
    value = eval(dev.pluginProps[u"conversionFormula"])
```

**Comment by HackintoshMan** [comment]
Yeah you would only need one formula for a device if you are only pulling one piece of data from the device. If you had:
```
# payload from the MQTT topic
payload = {
    "Time":"2019-08-04T09:16:14",
    "BME280":{
        "Temperature":84.3,
        "Humidity":31.3,
        "Pressure":995.74
    },
    "TempUnit":"F"
}
......
if dev.pluginProps[u"conversionFormula"]:
    p = payload
    payload = eval(dev.pluginProps[u"conversionFormula"])
```
Then, I think, you could even do something like `p["BME280"]["Temperature"] -= 12`

If you could get a multi-line text input, then you could even compile the multiple lines before feeding it into the `eval`.

Multi-line test from a config window (say stored in `dev.pluginProps[u"formula"]`):
```
p["BME280"]["Temperature"] -= 12
p["BME280"]["Humidity"] += 9
p["BME280"]["Pressure"] = p["BME280"]["Pressure"] / 33.8638
```

```
# payload from the MQTT topic
payload = {
    "Time":"2019-08-04T09:16:14",
    "BME280":{
        "Temperature":84.3,
        "Humidity":31.3,
        "Pressure":995.74
    },
    "TempUnit":"F"
}
......
if dev.pluginProps[u"formula"]:
    p = payload
    payload = eval(compile(dev.pluginProps[u"formula"]))
```

**Comment by FlyingDiver** [comment]
That seems overkill.  ;)

Unless you really wanted to adjust the values in the device states instead of just the device value.  Since you have multiple devices already, that would have to be done in each of them, which would be really confusing.

**Comment by HackintoshMan** [comment]
Right, that was the only thing I could come up with where you could modify multiple states if someone doesn't want to create a bunch of individual devices.

A single formula would do just fine for each device.

**Comment by FlyingDiver** [comment]
Implemented in 0.0.11
